dscache/capture_log.go

```diff
@@ -0,0 +1,96 @@
+package dscache
+
+import (
+	"context"
+	"fmt"
+
+	golog "github.com/ipfs/go-log"
+)
+
+// captureLog can replace this package's logger so that it counts calls instead of actually logging.
+// To be used only from tests
+type captureLog struct {
```

Comment:
I'm not a huge fan of using logs for tests. Why do you think this is the right approach in this context?

Reply:
So, I am unsure about this. The real goal of `TestConvertLogbookAndRefsMissingDsref` is to make sure that the dscache conversion will work if the logbook knows about a ref that is not found in the dsrefs. But this probably will never happen, while the opposite *will* happen since there are still cases were users (us developers) may have entries in our dsrefs that are not in logbook. Asserting that the log.Error in findFSIPath gets hit isn't really important to the test itself, but it helps ensure that expectations about the code are being met.

Could you elaborate on the problem you raised with using this mock log? Wondering if it's worth using it like this, or if the downside outweighs the (minor) benefit mentioned above.

Reply:
My concern is this change makes logging take on an additional meaning beyond just spitting out data about a running program. Adding or removing logging information should never break a program (tests are a program), and that'll happen here if a developer alters logging output.

We haven't yet made the jump from just logging stuff to gathering observability metrics which is the thing we _should_ be overloading logging with, but when we do I expect our approach to logging to change a bunch. I'd heavily prefer changes to logging _never_ affect our test suite.

If we need to test something that feels like it can't be tested, let's change code to make it testable.

Reply:
Okay, removed. While exploring ahead I realized this function `findFSIPath` is going to change anyway, so it's also not worth adding here.